File: fastapi/routes/books.py
from fastapi import APIRouter, Depends, Query, HTTPException
from sqlalchemy.orm import Session

from pydantic import BaseModel, Field, conlist
from typing import Optional, List, Literal, Dict, Any

from services.book_service import BookService
from core.dependencies import get_db, get_current_user
from models.book import Book

router = APIRouter(prefix="/api/v1/books", tags=["books"])

# -------- Helper: Convert to frontend format --------
def _to_read(b: Book) -> dict:
    return {
        "id": b.id,
        "ownerId": b.owner_id,
        "titleOr": b.title_or,
        "titleEn": b.title_en,
        "originalLanguage": b.original_language,
        "author": b.author,
        "category": b.category,
        "description": b.description,
        "coverImgUrl": b.cover_img_url,
        "conditionImgURLs": b.condition_img_urls or [],
        "status": b.status,
        "condition": b.condition,
        "canRent": b.can_rent,
        "canSell": b.can_sell,
        "dateAdded": b.date_added,
        "updateDate": b.update_date,
        "isbn": b.isbn,
        "tags": b.tags or [],
        "publishYear": b.publish_year,
        "maxLendingDays": b.max_lending_days,
        "deliveryMethod": b.delivery_method,
        "salePrice": b.sale_price,
        "deposit": b.deposit,
    }


class BookCreate(BaseModel):
    titleOr: str
    titleEn: Optional[str] = None
    originalLanguage: Optional[str] = None
    author: str
    category: Optional[str] = None
    description: Optional[str] = ""
    coverImgUrl: Optional[str] = None
    conditionImgURLs: Optional[List[str]] = None
    status: Literal["listed","unlisted","lent","sold"] = "listed"
    condition: Literal["new","like-new","good","fair"] = "good"
    canRent: bool = True
    canSell: bool = False
    isbn: Optional[str] = None
    tags: Optional[List[str]] = None
    publishYear: Optional[int] = None
    maxLendingDays: int = 14
    deliveryMethod: Literal["post","pickup","both"] = "both"
    salePrice: Optional[float] = None
    deposit: Optional[float] = None

class BookUpdate(BaseModel):
    titleOr: Optional[str] = None
    titleEn: Optional[str] = None
    originalLanguage: Optional[str] = None
    author: Optional[str] = None
    category: Optional[str] = None
    description: Optional[str] = None
    coverImgUrl: Optional[str] = None
    conditionImgURLs: Optional[List[str]] = None
    status: Optional[Literal["listed","unlisted","lent","sold"]] = None
    condition: Optional[Literal["new","like-new","good","fair"]] = None
    canRent: Optional[bool] = None
    canSell: Optional[bool] = None
    isbn: Optional[str] = None
    tags: Optional[List[str]] = None
    publishYear: Optional[int] = None
    maxLendingDays: Optional[int] = None
    deliveryMethod: Optional[Literal["post","pickup","both"]] = None
    salePrice: Optional[float] = None
    deposit: Optional[float] = None


# -------- Create --------
@router.post("", response_model=dict)
def create_book(
    payload: BookCreate,
    db: Session = Depends(get_db),
    user = Depends(get_current_user),
):
    p = {
        "title_or": payload.titleOr,
        "title_en": payload.titleEn,
        "original_language": payload.originalLanguage,
        "author": payload.author,
        "category": payload.category,
        "description": payload.description or "",
        "cover_img_url": payload.coverImgUrl,
        "condition_img_urls": payload.conditionImgURLs,
        "status": payload.status,
        "condition": payload.condition,
        "can_rent": payload.canRent,
        "can_sell": payload.canSell,
        "isbn": payload.isbn,
        "tags": payload.tags,
        "publish_year": payload.publishYear,
        "max_lending_days": payload.maxLendingDays,
        "delivery_method": payload.deliveryMethod,
        "sale_price": payload.salePrice,
        "deposit": payload.deposit,
    }
    book = BookService.create(db, owner_id=user.user_id, payload=p)
    return _to_read(book)


# -------- List --------
@router.get("", response_model=dict)
def list_books(
    page: int = Query(1, ge=1),
    page_size: int = Query(20, ge=1, le=100),
    q: Optional[str] = None,
    author: Optional[str] = None,
    category: Optional[str] = None,
    owner_id: Optional[str] = None,
    status: Optional[str] = None,
    db: Session = Depends(get_db),
):
    items, total = BookService.list(db, page, page_size, q, author, category, owner_id, status)
    return {"items": [_to_read(b) for b in items], "total": total, "page": page, "page_size": page_size}


# -------- Get Filter Options --------
@router.get("/filter-options")
def get_filter_options(db: Session = Depends(get_db)):
    """Get all unique categories and languages for filter dropdowns."""
    return BookService.get_filter_options(db)


# -------- Search --------
@router.get("/search")
def search_books(
    q: Optional[str] = Query(None, description="keyword in title/author/description/ISBN"),
    lang: Optional[str] = Query(None),
    category: Optional[str] = None,
    status: Literal["listed","unlisted","all"] = "listed",
    canRent: Optional[bool] = None,
    canSell: Optional[bool] = None,
    delivery: Literal["any","post","pickup","both"] = "any",
    minPrice: Optional[float] = None,
    maxPrice: Optional[float] = None,
    sort: Literal["relevance","newest","price_asc","price_desc"] = "relevance",
    page: int = 1,
    page_size: int = 20,
    db: Session = Depends(get_db),
):
    items, total = BookService.search_books(
        db=db, q=q, lang=lang, category=category, status=status, can_rent=canRent, can_sell=canSell,
        delivery=delivery, min_price=minPrice, max_price=maxPrice,
        sort=sort, page=page, page_size=page_size
    )
    return {"items": items, "total": total, "page": page, "page_size": page_size}


# -------- Get by ID --------
@router.get("/{book_id}", response_model=dict)
def get_book(book_id: str, db: Session = Depends(get_db)):
    return _to_read(BookService.get(db, book_id))

# -------- Update --------
@router.put("/{book_id}", response_model=dict)
def update_book(
    book_id: str,
    payload: BookUpdate,
    db: Session = Depends(get_db),
    user = Depends(get_current_user),
):
    mapping = {
        "titleOr":"title_or","titleEn":"title_en","originalLanguage":"original_language",
        "coverImgUrl":"cover_img_url","conditionImgURLs":"condition_img_urls",
        "canRent":"can_rent","canSell":"can_sell",
        "publishYear":"publish_year","maxLendingDays":"max_lending_days",
        "deliveryMethod":"delivery_method","salePrice":"sale_price"
    }
    data = {mapping.get(k, k): v for k, v in payload.dict(exclude_unset=True).items()}
    book = BookService.update(db, book_id, user.user_id, data)
    return _to_read(book)


# -------- Delete --------
@router.delete("/{book_id}", status_code=204)
def delete_book(
    book_id: str,
    db: Session = Depends(get_db),
    user = Depends(get_current_user),
):
    BookService.delete(db, book_id, user.user_id)
    return None

File: frontendNext/utils/books.ts
import axios from "axios";
import { getToken, getApiUrl } from "./auth";
import type { Book } from "@/app/types/book";

// Get filter options (all categories and languages from database)
export const getFilterOptions = async (): Promise<{
  categories: string[];
  languages: string[];
}> => {
  const API_URL = getApiUrl();

  try {
    const response = await axios.get(`${API_URL}/api/v1/books/filter-options`);
    return response.data;
  } catch (error) {
    console.error("Failed to fetch filter options:", error);
    return { categories: [], languages: [] };
  }
};

// start lending - save a new book
export const createBook = async (book: Book) => {
  const API_URL = getApiUrl();

  try {
    const response = await axios.post(
      `${API_URL}/api/v1/books`,
      book,
      {
        headers: {
          "Content-Type": "application/json",
          Authorization: `Bearer ${localStorage.getItem("access_token")}`,
        },
        withCredentials: true,
      }
    );

    return response.data;
  } catch (error) {
    console.error("Create Book API (JSON) failed:", error);
    throw error;
  }
};

// Get books（ owner / status / search / paging）
export const getBooks = async (params?: {
  ownerId?: string;
  status?: string;   // listed | unlisted | lent | sold
  q?: string;
  page?: number;
  pageSize?: number;
}): Promise<Book[]> => {
  const API_URL = getApiUrl();
  const token = getToken();

  try {
    const response = await axios.get(`${API_URL}/api/v1/books`, {
      // Attach token only if it exists
      // If user is logged in -> send Authorization header
      // If not logged in -> request without token (guest access)
      headers: token ? { Authorization: `Bearer ${token}` } : {},

      params: {
        owner_id: params?.ownerId,
        status: params?.status,
        q: params?.q,
        page: params?.page ?? 1,
        page_size: Math.min(params?.pageSize ?? 20, 100),
      },
    });

    const data = response.data;
    const raw: any[] = Array.isArray(data)
      ? data
      : Array.isArray(data.items)
      ? data.items
      : Array.isArray(data.data)
      ? data.data
      : [];

    return raw as Book[];
  } catch (error) {
    console.error("Get Books API failed:", error);
    return [];
  }
};

/** Get a book by id */
export const getBookById = async (bookId: string): Promise<Book | null> => {
  const API_URL = getApiUrl();
  const token = getToken();

  try {
    const res = await axios.get(`${API_URL}/api/v1/books/${bookId}`, {
      headers: token ? { Authorization: `Bearer ${token}` } : {},
    });
      console.log("Response status:", res.status);

    return res.data as Book;
  } catch (err) {
    console.error("Get Book By Id failed:", err);
    return null;
  }
};

// Upload images
export async function uploadFile (file: File, scene: string): Promise<string> {
  const API_URL = getApiUrl();
  const token = getToken();
    if (!scene || !scene.trim()) throw new Error("scene is required");


  const formData = new FormData();
  formData.append("file", file);
  formData.append("scene", scene);

  const res = await axios.post(`${API_URL}/upload/image`, formData, {
  headers: {
    "Content-Type": "multipart/form-data",
    Authorization: `Bearer ${token}`,
  },
});


  // return a URL
  return `${API_URL}${res.data.path}`; 
}

// Update book
export async function updateBook(bookId: string, payload: Partial<Book>) {
  const API_URL = getApiUrl();
  const res = await axios.put(`${API_URL}/api/v1/books/${bookId}`, payload, {
    headers: {
      "Content-Type": "application/json",
      Authorization: `Bearer ${getToken()}`,
    },
  });
  return res.data;
}

// delete book
export const deleteBook = async (bookId: string): Promise<boolean> => {
  const API_URL = getApiUrl();
  const token = getToken();
  if (!token) return false;

  try {
    await axios.delete(`${API_URL}/api/v1/books/${bookId}`, {
      headers: {
        Authorization: `Bearer ${token}`,
      },
    });
    return true;
  } catch (error) {
    console.error("Delete Book API failed:", error);
    return false;
  }
};

export interface SearchParams {
  q?: string;
  lang?: string;
  category?: string;
  status?: 'listed' | 'unlisted' | 'lent' | 'sold';
  canRent?: boolean;
  canSell?: boolean;
  delivery?: 'post' | 'pickup' | 'both' | 'any';
  minPrice?: number;
  maxPrice?: number;
  sort?: 'relevance' | 'newest' | 'price_asc' | 'price_desc';
  page?: number;
  pageSize?: number;
}

export async function searchBooks(params: SearchParams) {
  const API_URL = getApiUrl();
  const token = getToken();
  
  const searchParams = new URLSearchParams();
  
  // Add non-null parameters to query string
  Object.entries(params).forEach(([key, value]) => {
    if (value !== undefined && value !== null) {
      searchParams.append(key, value.toString());
    }
  });

  const response = await axios.get(`${API_URL}/api/v1/books/search?${searchParams}`);
  const data = response.data;

  const normalizedItems = data.items.map((b: any) => ({
    id: b.id,
    titleOr: b.titleOr,
    author: b.author,
    coverImgUrl: b.coverImgUrl,
    deliveryMethod: b.deliveryMethod,
    dateAdded: b.dateAdded || b.createdAt,
    ownerId: b.ownerId,
    canRent: b.canRent,
    canSell: b.canSell,
    deposit: b.deposit,
    salePrice: b.salePrice,
    status: b.status,
    tags: b.tags || [],
    category: b.category,
    originalLanguage: b.originalLanguage || b.lang,
  }));

  return { items: normalizedItems, total: data.total };
};